packages/core/src/proxy_configuration.ts

```diff
@@ -189,6 +265,52 @@ export class ProxyConfiguration {
         };
     }
 
+    /**
+     * Given a session identifier and a request / proxy tier, this function returns a new proxy URL based on the provided configuration options.
+     * @param _sessionId Session identifier
+     * @param options Options for the tiered proxy rotation
+     * @returns A string with a proxy URL.
+     */
+    protected _handleTieredUrl(_sessionId: string, options?: TieredProxyOptions): string {
+        if (!this.tieredProxyUrls) throw new Error('Tiered proxy URLs are not set');
+
+        if (!options || (!options?.request && options?.proxyTier === undefined)) {
+            const allProxyUrls = this.tieredProxyUrls.flat();
+            return allProxyUrls[this.nextCustomUrlIndex++ % allProxyUrls.length];
+        }
+
+        let tierPrediction = options.proxyTier!;
+
+        if (typeof tierPrediction !== 'number') {
+            tierPrediction = this.getProxyTier(options.request!)!;
+        }
+
+        return this.tieredProxyUrls![tierPrediction][this.nextCustomUrlIndex++ % this.tieredProxyUrls![tierPrediction].length];
+    }
+
+    /**
+     * Given a `Request` object, this function returns the tier of the proxy that should be used for the request.
+     *
+     * This returns `null` if `tieredProxyUrls` option is not set.
+     */
+    getProxyTier(request: Request): number | null {
+        if (!this.tieredProxyUrls) return null;
+
+        const domain = new URL(request.url).hostname;
+        if (!this.domainTiers.has(domain)) {
+            this.domainTiers.set(domain, new ProxyTierTracker(this.tieredProxyUrls));
+        }
+
+        const tracker = this.domainTiers.get(domain)!;
+        const tierPrediction = tracker.getTier();
+
+        request.addHook('sessionRotation', () => {
+            tracker.addError(tierPrediction);
+        });
```

Comment:
I wouldn't expect a function called `getProxyTier` to modify its arguments in any way.

Reply:
Yeah, good point - the new approach still does this, but in a more... manageable way? I'm curious to hear your opinion there.